### PR TITLE
enhance: add message and msgstream msgpack adaptor

### PR DESCRIPTION
**Pull request by chyezh**
issue: #33285

- make message builder and message conversion type safe
- add adaptor type and function to adapt old msgstream msgpack and message interface

### CODE REVIEW

File: pkg/util/typeutil/multipart_queue.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package typeutil
```
Comment (chyezh) [inline pkg/util/typeutil/multipart_queue.go:1]: move from wal.utility  for common usage.

File: pkg/streaming/walimpls/impls/pulsar/message_id.go [hunk 2]
```diff
@@ -61,13 +61,13 @@ func (id pulsarID) LTE(other message.MessageID) bool {
...
+	id2 := other.(PulsarID)
 	return id.LedgerID() == id2.LedgerID() &&
 		id.EntryID() == id2.EntryID() &&
 		id.BatchIdx() == id2.BatchIdx()
 }
 
-func (id pulsarID) Marshal() string {
-	return hex.EncodeToString(id.Serialize())
+func (id PulsarID) Marshal() string {
+	return base64.StdEncoding.EncodeToString(id.Serialize())
```
Comment (chyezh) [inline pkg/streaming/walimpls/impls/pulsar/message_id.go:72]: use base64 for smaller binary representation

File: pkg/streaming/walimpls/impls/pulsar/message_id.go [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
+	"encoding/base64"
 
 	"github.com/apache/pulsar-client-go/pulsar"
 	"github.com/cockroachdb/errors"
 
 	"github.com/milvus-io/milvus/pkg/streaming/util/message"
 )
 
-var _ message.MessageID = pulsarID{}
+var _ message.MessageID = PulsarID{}
```
Comment (chyezh) [inline pkg/streaming/walimpls/impls/pulsar/message_id.go:12]: make type public for adaption.

File: internal/streamingnode/server/wal/adaptor/message_handler.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	return &MsgPackAdaptorHandler{
+		logger:         log.With(),
+		channel:        make(chan *msgstream.MsgPack),
+		pendings:       make([]message.ImmutableMessage, 0),
+		pendingMsgPack: typeutil.NewMultipartQueue[*msgstream.MsgPack](),
+	}
+}
+
+// MsgPackAdaptorHandler is the handler for message pack.
+type MsgPackAdaptorHandler struct {
```
Comment (chyezh) [inline internal/streamingnode/server/wal/adaptor/message_handler.go:50]: add message pack adpator to adapt message into msgpack

File: pkg/streaming/util/message/builder.go [hunk 5]
```diff
@@ -15,43 +33,69 @@ func NewImmutableMesasge(
 	}
 }
 
-// NewMutableMessageBuilder creates a new builder.
+// List all type-safe mutable message builders here.
+var (
+	NewTimeTickMessageBuilderV1         = createNewMessageBuilderV1[*TimeTickMessageHeader, *msgpb.TimeTickMsg]()
```
Comment (chyezh) [inline pkg/streaming/util/message/builder.go:38]: type safe builder for message type

File: pkg/streaming/util/message/specialized_message.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	reflect.TypeOf(&DeleteMessageHeader{}):           MessageTypeDelete,
+	reflect.TypeOf(&CreateCollectionMessageHeader{}): MessageTypeCreateCollection,
+	reflect.TypeOf(&DropCollectionMessageHeader{}):   MessageTypeDropCollection,
+	reflect.TypeOf(&CreatePartitionMessageHeader{}):  MessageTypeCreatePartition,
+	reflect.TypeOf(&DropPartitionMessageHeader{}):    MessageTypeDropPartition,
+}
+
+// List all specialized message types.
+type (
+	MutableTimeTickMessage  = specializedMutableMessage[*TimeTickMessageHeader]
```
Comment (chyezh) [inline pkg/streaming/util/message/specialized_message.go:37]: type safe conversion for message type

File: pkg/streaming/util/message/adaptor/message.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		switch msg.Version() {
+		case message.VersionOld:
+			tsMsgs, err = fromMessageToTsMsgVOld(msg)
+		case message.VersionV1:
+			tsMsgs, err = fromMessageToTsMsgV1(msg)
+		default:
+			panic("unsupported message version")
+		}
+		if err != nil {
+			finalErr = errors.CombineErrors(finalErr, errors.Wrapf(err, "Failed to convert message to msgpack, %v", msg.MessageID()))
```
Comment (jaime0815) [inline pkg/streaming/util/message/adaptor/message.go:43]: Return immediately if the error is not nil; checking len(allTsMsgs) == 0 may skip some illegal messages at line 48.

File: pkg/streaming/util/message/adaptor/message.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		panic("unreachable code, partition id is not exist")
+	}
+	tsMsgs := make([]msgstream.TsMsg, 0, len(assignments.SegmentAssignments))
+	// timetick should has been assign at streaming node.
+	// so overwrite the timetick on insertRequest.
+
+	left := 0
+	right := 0
+	for _, assignment := range assignments.SegmentAssignments {
+		right = left + int(assignment.Row)
```
Comment (jaime0815) [inline pkg/streaming/util/message/adaptor/message.go:127]: Wrap an independent method to build `tsMsgs`  slice for a `segmentAssignment`.

File: pkg/streaming/util/message/adaptor/message.go (hunk 7 above)
Comment (chyezh) [reply pkg/streaming/util/message/adaptor/message.go:43]: no, we only filter out illegal message, but keep the remaining legal message to consume.

File: pkg/streaming/util/message/adaptor/message.go (hunk 8 above)
Comment (chyezh) [reply pkg/streaming/util/message/adaptor/message.go:127]: ok, fix it soon

File: pkg/streaming/util/message/adaptor/message.go (hunk 7 above)
Comment (jaime0815) [reply pkg/streaming/util/message/adaptor/message.go:37]: No illegal messages are expected. Log or return an error when filtering out illegal messages; only return an error if all messages are illegal. It is confusing that you shouldn't return an error on line 43 when all messages are illegal if the aim is to skip those messages?

**Pull request merged**